### PR TITLE
feat: Bundle Server dependencies

### PR DESCRIPTION
**Pull request by AshCorr**
## What does this change?

 - Uses Webpack to bundle our Server dependencies which lets us tree-shake a lot of unused code in our dependencies away.
 - Replaces `html-minifier` with its fork `html-minifier-terser`. `html-minifier` does not play nicely with Webpack due to its dependency on UglifyJS, meanwhile `html-minifier-terser` uses terser which works fine in Webpack (I believe its also the library Webpack uses by default for minifying JS?)
## Why?

This allows users and actions to download and run our container images faster. It probably speeds up deployments a bit too, but likely only by a minimal amount, and it won't be noticed anyway due to how we wait 60 seconds for deployments to warm up anyway.

|       Type                   |   Before  | After | Reduction |
|------------------------------|-----------|-------|-----------|
| Container Image (compressed) | 361MB     | 56MB  |    85%    |
| rendering.zip                | 400MB     | 12MB  |    97%    |

## Build times (test on my Mac)

Running `time make build`

Before: 109.61s 102.28s 103.73s
After: 116.93s 122.94s 121.46s

Running `time make riffraff-bundle`

Before: 142.59s 139.38s 136.92s
After: 123.14s 126.19s 128.63s

Publishing image to GHCR

Before: 34s
After: 10s

The webpack build takes significantly longer with these changes, which is what we'd expect as it has to do a lot more work to build the server bundle now, **however** we make up all that time by having far less data to compress when creating the final Riffraff ZIP bundle.

## Impact on dev server

EDIT: Just realised the dev server uses the same config, I was hoping with getting away with only making these changes for the prod server. Need to investigate impact.

EDIT 2: Adds 1.5 sec to dev server startup, should be able to have it revert to the old behaviour in dev mode though.

### CODE REVIEW

File: dotcom-rendering/src/amp/components/Elements.tsx [hunk 1]
```diff
@@ -109,8 +108,7 @@ export const Elements = (
 	isImmersive: boolean,
 	adTargeting?: AdTargeting,
 ): JSX.Element[] => {
-	const cleanedElements = enhance(elements);
```
Comment (AshCorr) [inline dotcom-rendering/src/amp/components/Elements.tsx:112]: I've pulled this up into AMPs `index.tsx`, the new minifer library uses async which is hard to use in Components.

File: dotcom-rendering/src/amp/lib/enhance.ts [hunk 2]
```diff
@@ -7,28 +7,32 @@ import type { FEElement } from '../../types/content';
...
-const clean = (html: string) => {
+const clean = async (html: string) => {
 	const sanitised = sanitiseHTML(html, {
 		ADD_TAGS: ['#comment'],
 		FORCE_BODY: true,
 	});
 	const transformed = bigBullets(sanitised);
 
-	return minify(transformed, {
+	return await minify(transformed, {
```
Comment (arelra) [inline dotcom-rendering/src/amp/lib/enhance.ts:17]: I _think_ you don't need the async or await here as it returns a promise which gets awaited separately in `enhance`?

**Review by arelra — COMMENTED** [review]
Great to see this!

Fyi I tried `make dev` locally and I'm getting a runtime  error:

Screenshot 2023-04-03 at 11 30 17

**Comment by AshCorr** [comment]
I believe Express and Log4js are trying to dynamically import dependencies and webpack doesn't know how to deal with it. For example:

```
function doesAThing() {
   var something = import("another thing");
}
```

In expresses case its trying to see if there are any template rendering libraries and load them, since we don't use that functionality of Express we're fine.

Log4js is a little different though, its looks like its trying to load an appender library, this is something that we might need for our ELK logging to work correctly, need to investigate further.

**Comment by OllysCoding** [comment]
> @AshCorr Log4js is a little different though, it's looks like its trying to load an appender library, this is something that we might need for our ELK logging to work correctly, need to investigate further.

Can you somehow manually tell webpack to bundle this library in the webpack server config, then the dynamic import would (hopefully) just work? Not sure exactly what you would need to do in the webpack config to tell it to bundle that appender library but I'm sure there's a way ...

**Comment by AshCorr** [comment]
> Can you somehow manually tell webpack to bundle this library in the webpack server config, then the dynamic import would (hopefully) just work? Not sure exactly what you would need to do in the webpack config to tell it to bundle that appender library but I'm sure there's a way ...

It looks like we're only using the `file` and `console` appenders from Log4js which are considered as "Core appenders" and always imported, so I think we're actually fine. Still, I'd love to try and figure out if theres a way we can get rid of the warning without just adding an ignore for it.

**Review by OllysCoding — APPROVED** [review]
+1 This is excellent work!

From my testing this increases the dev-build server build time from 4.5s to 6.5s (on my machine), I was worried because pre-SWC the differences was closer to 15-30 seconds see https://github.com/guardian/dotcom-rendering/pull/3906) last year!

I think a 2s increase is completely acceptable, but if we did want to, we could externalise node_modules only on dev! 

An interesting related question to the above is whether we have done any rendering time tests on the externalised node_modules vs bundled? Testing on my machine made me wonder if this new bundled approach is a little bit faster (at least on dev?)

**Comment by OllysCoding** [comment]
Oh and I had meant to add we should definitely test this on CODE before merging!

**Comment by AshCorr** [comment]
Tested in CODE, all seems good!

File: dotcom-rendering/scripts/deploy/build-riffraff-bundle.js [hunk 3]
```diff
@@ -67,30 +67,41 @@ const copyDist = () => {
...
-			maxBuffer: 200000000, // increase if you get a maxBuffer exceeded error
-		},
-	).then(() => {
-		cpy(['rendering.zip'], path.resolve(target, 'rendering', 'dist'));
+	return execa('zip', ['--move', '--recurse-paths', 'rendering.zip', '.'], {
+		cwd: path.resolve(target, 'rendering'),
 	});
 };
 
-const removeAR = () => {
```
Comment (AshCorr) [inline dotcom-rendering/scripts/deploy/build-riffraff-bundle.js:89]: We no longer delete the AR folder as we're a lot more specific about what we zip into our production bundle

File: dotcom-rendering/cloudformation.yml [hunk 4]
```diff
@@ -247,12 +247,12 @@ Resources:
           usermod -a -G frontend aws-kinesis-agent-user
           cd /home/dotcom-rendering
 
-          aws --region eu-west-1 s3 cp s3://aws-frontend-artifacts/frontend/${Stage}/${App}/dist/${App}.zip ./
+          aws --region eu-west-1 s3 cp s3://aws-frontend-artifacts/frontend/${Stage}/${App}/${App}.zip ./
           unzip -q ${App}.zip -d ${App}
 
           chown -R dotcom-rendering:frontend ${App}
 
-          cd ${App}/dotcom-rendering
```
Comment (AshCorr) [inline dotcom-rendering/cloudformation.yml:255]: We're a lot more specific when creating the ZIP bundle now and everything required to run the DCR server is in the root folder of the bundle now.

Previously the bundle contained the entire repository, including common-rendering, which is why we'd have to cd into the dotcom-rendering subfolder.

**Pull request merged**